From 63b9e2aa29c5e6402aaeb973b938b301e3791fe0 Mon Sep 17 00:00:00 2001
From: Aloka Dixit <quic_alokad@quicinc.com>
Date: Tue, 21 Sep 2021 13:23:28 -0700
Subject: [PATCH 16/27] ath12k: EHT operation element support in AP mode

Pass EHT operation element parameters to target during AP bring up
and during peer association.

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c | 18 ++++++++++++++++++
 drivers/net/wireless/ath/ath12k/wmi.c |  5 +++--
 drivers/net/wireless/ath/ath12k/wmi.h | 10 ++++++++++
 3 files changed, 31 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index 7377787a7fc7..ad0a90b330d6 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -2072,6 +2072,8 @@ static void ath12k_peer_assoc_h_eht(struct ath12k *ar,
 	       sizeof(eht_cap->eht_cap_elem.mac_cap_info));
 	memcpy(arg->peer_eht_cap_phy, eht_cap->eht_cap_elem.phy_cap_info,
 	       sizeof(eht_cap->eht_cap_elem.phy_cap_info));
+	memcpy(&arg->peer_eht_ops, &vif->bss_conf.eht_oper,
+	       sizeof(vif->bss_conf.eht_oper));
 
 	switch (sta->bandwidth) {
 	case IEEE80211_STA_RX_BW_320:
@@ -2924,6 +2926,22 @@ static void ath12k_mac_op_bss_info_changed(struct ieee80211_hw *hw,
 			if (ret)
 				ath12k_warn(ar->ab, "Failed to set he oper params %x for VDEV %d: %i\n",
 					    param_value, arvif->vdev_id, ret);
+
+			param_id = WMI_VDEV_PARAM_EHTOPS_0_31;
+			memcpy(&param_value, &vif->bss_conf.eht_oper,
+			       sizeof(vif->bss_conf.eht_oper));
+			ret = ath12k_wmi_vdev_set_param_cmd(ar,
+					arvif->vdev_id,
+					param_id,
+					param_value);
+			ath12k_dbg(ar->ab, ATH12K_DBG_MAC,
+					"eht oper param: %x set for VDEV: %d\n",
+					param_value, arvif->vdev_id);
+
+			if (ret)
+				ath12k_warn(ar->ab,
+					    "Failed to set eht oper params %x for VDEV %d: %i\n",
+					    param_value, arvif->vdev_id, ret);
 		}
 	}
 
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index 4a4c552fd358..f4ee1ce7e885 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -1933,6 +1933,7 @@ int ath12k_wmi_send_peer_assoc_cmd(struct ath12k *ar,
 		       0);
 	memcpy_and_pad(&cmd->peer_eht_ppet, sizeof(cmd->peer_eht_ppet),
 		       &param->peer_eht_ppet, sizeof(param->peer_eht_ppet), 0);
+	cmd->peer_eht_ops = param->peer_eht_ops;
 
 	/* Update peer legacy rate information */
 	ptr += sizeof(*cmd);
@@ -2046,7 +2047,7 @@ int ath12k_wmi_send_peer_assoc_cmd(struct ath12k *ar,
 	}
 
 	ath12k_dbg(ar->ab, ATH12K_DBG_WMI,
-		   "wmi peer assoc vdev id %d assoc id %d peer mac %pM peer_flags %x rate_caps %x peer_caps %x listen_intval %d ht_caps %x max_mpdu %d nss %d phymode %d peer_mpdu_density %d vht_caps %x he cap_info %x he ops %x he cap_info_ext %x he phy %x %x %x peer_bw_rxnss_override %x peer_flags_ext %x eht mac_cap %x %x eht phy_cap %x %x %x\n",
+		   "wmi peer assoc vdev id %d assoc id %d peer mac %pM peer_flags %x rate_caps %x peer_caps %x listen_intval %d ht_caps %x max_mpdu %d nss %d phymode %d peer_mpdu_density %d vht_caps %x he cap_info %x he ops %x he cap_info_ext %x he phy %x %x %x peer_bw_rxnss_override %x peer_flags_ext %x eht mac_cap %x %x eht phy_cap %x %x %x eht ops %x\n",
 		   cmd->vdev_id, cmd->peer_associd, param->peer_mac,
 		   cmd->peer_flags, cmd->peer_rate_caps, cmd->peer_caps,
 		   cmd->peer_listen_intval, cmd->peer_ht_caps,
@@ -2059,7 +2060,7 @@ int ath12k_wmi_send_peer_assoc_cmd(struct ath12k *ar,
 		   cmd->peer_bw_rxnss_override, cmd->peer_flags_ext,
 		   cmd->peer_eht_cap_mac[0], cmd->peer_eht_cap_mac[1],
 		   cmd->peer_eht_cap_phy[0], cmd->peer_eht_cap_phy[1],
-		   cmd->peer_eht_cap_phy[2]);
+		   cmd->peer_eht_cap_phy[2], cmd->peer_eht_ops);
 
 	return ret;
 }
diff --git a/drivers/net/wireless/ath/ath12k/wmi.h b/drivers/net/wireless/ath/ath12k/wmi.h
index 06063ff7362e..6429daaff182 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -1058,6 +1058,15 @@ enum wmi_tlv_vdev_param {
 	WMI_VDEV_PARAM_BSS_COLOR,
 	WMI_VDEV_PARAM_SET_HEMU_MODE,
 	WMI_VDEV_PARAM_HEOPS_0_31 = 0x8003,
+	WMI_VDEV_PARAM_OBSSPD,
+	WMI_VDEV_PARAM_SET_EHT_MU_MODE,
+	WMI_VDEV_PARAM_EHT_LTF,
+	WMI_VDEV_PARAM_UL_EHT_LTF,
+	WMI_VDEV_PARAM_EHT_DCM,
+	WMI_VDEV_PARAM_EHT_RANGE_EXT,
+	WMI_VDEV_PARAM_NON_DATA_EHT_RANGE_EXT,
+	WMI_VDEV_PARAM_FIXED_PUNCTURE_PATTERN,
+	WMI_VDEV_PARAM_EHTOPS_0_31,
 };
 
 enum wmi_tlv_peer_flags {
@@ -3769,6 +3778,7 @@ struct peer_assoc_params {
 	u32 peer_eht_mcs_count;
 	u32 peer_eht_rx_mcs_set[WMI_MAX_EHTCAP_RATE_SET];
 	u32 peer_eht_tx_mcs_set[WMI_MAX_EHTCAP_RATE_SET];
+	u32 peer_eht_ops;
 	struct ath12k_ppe_threshold peer_eht_ppet;
 };
 
-- 
2.31.1

